test/test_get_ops.js

```diff
@@ -0,0 +1,153 @@
+var expect = require('chai').expect;
+var ShareDbMongo = require('..');
+var getQuery = require('sharedb-mingo-memory/get-query');
+var sinon = require('sinon');
+
+var mongoUrl = process.env.TEST_MONGO_URL || 'mongodb://localhost:27017/test';
+
+function create(callback) {
+  var db = new ShareDbMongo(mongoUrl, {
+    mongoOptions: {},
+    getOpsWithoutStrictLinking: true
+  });
+  db.getDbs(function(err, mongo) {
+    if (err) return callback(err);
+    mongo.dropDatabase(function(err) {
+      if (err) return callback(err);
+      callback(null, db, mongo);
+    });
+  });
+};
+
+require('sharedb/test/db')({create: create, getQuery: getQuery});
+
+describe('getOps', function() {
+  beforeEach(function(done) {
+    var self = this;
+    create(function(err, db, mongo) {
+      if (err) return done(err);
+      self.db = db;
+      self.mongo = mongo;
+      done();
+    });
+  });
+
+  afterEach(function(done) {
+    this.db.close(done);
+  });
+
+  describe('a chain of ops', function() {
+    var db;
+    var mongo;
+    var id;
+    var collection;
+
+    beforeEach(function(done) {
+      db = this.db;
+      mongo = this.mongo;
+      id = 'document1';
+      collection = 'testcollection';
+
+      sinon.spy(db, '_getOps');
+      sinon.spy(db, '_getSnapshotOpLink');
+
+      var ops = [
+        {v: 0, create: {}},
+        {v: 1, p: ['foo'], oi: 'bar'},
+        {v: 2, p: ['foo'], oi: 'baz'},
+        {v: 3, p: ['foo'], oi: 'qux'}
+      ];
+
+      commitOpChain(db, mongo, collection, id, ops, function(error) {
+        if (error) done(error);
+        mongo.collection('o_' + collection).deleteOne({v: 1}, done);
+      });
+    });
+
+    it('fetches ops 2-3 without fetching all ops', function(done) {
+      db.getOps(collection, id, 2, 4, null, function(error, ops) {
+        if (error) return done(error);
+        expect(ops.length).to.equal(2);
+        expect(ops[0].v).to.equal(2);
+        expect(ops[1].v).to.equal(3);
+        done();
+      });
+    });
+
+    it('default option errors when missing ops', function(done) {
+      db.getOps(collection, id, 0, 4, null, function(error) {
+        expect(error.code).to.equal(5103);
+        expect(error.message).to.equal('Missing ops from requested version testcollection.document1 0');
+        done();
+      });
+    });
+
+    it('ignoreMissingOps option returns ops up to the first missing op', function(done) {
```

Comment:
This is worth talking about. I haven't tried this locally, but I'm pretty sure this behaviour only happens because you've set `getOpsWithoutStrictLinking: true`, which tries to iterate **forward** through ops.

However, the **default** behaviour (`getOpsWithoutStrictLinking: false`) is to start from the snapshot, and then iterate **backwards** through ops.

This means we get inconsistent behaviour, depending on whether `getOpsWithoutStrictLinking` is enabled or not. If I have an op missing in the middle of my range, then:

 - `getOpsWithoutStrictLinking: false` will return the **end** of the range
 - `getOpsWithoutStrictLinking: true` will return the **start* of the range

One other point on `getOpsWithoutStrictLinking` (I think as we discussed very briefly yesterday), using `getOpsWithoutStrictLinking` on a collection that has been TTLed has slightly questionable validity. It's built around the assumption that if you can find an op with a unique version, then that op can be considered canonical. However, in the case of TTL, your TTL mechanism may or may not have deleted other ops with the same version, leaving a non-canonical op behind. The odds of this happening are probably pretty low, but it's still possible, and something to consider in a collection where ops are known to have been deleted.

I wonder if you set `ignoreMissingOps: true` if we should just _ignore_ `getOpsWithoutStrictLinking`? Then we'd have consistent behaviour, at least. Also, as discussed above, this is the surest way to make sure you're returning canonical ops, since the `getOpsWithoutStrictLinking` assumptions may be broken.

Reply:
Might be the phrasing of the test description it issue, but it does appear to work backwards until the first missing op in both cases. So when given a set of ops `[0,2,3]` and fetching `0..4` it gets `[2,3]` stopping at `1` and doesn't return `0` in either case.

Reply:
Oh yes! I should probably read the test case code better 😅 

Sorry I got confused, because I remembered `getOpsWithStrictLinking` iterating forwards, which it does, but it does it from the **end** of the requested op range, and _then_ iterates backwards again, which _is_ consistent with the default behaviour.

At any rate, it may be worth adding a test case with `getOpsWithoutStrictLinking` disabled, just to make sure we don't break this consistency in future.